### DISCUSSION TITLE
Integrate with Discord

### DISCUSSION THREAD

**Discussion by milivojm** [CntoDev/level-up#88] [General]
When Member gets promoted, assign appropriate tags on Discord.

`MemberPromoted` event should be listened.

**Comment by milivojm** [comment]
Alright, I think I've figured it out. So, we have three systems:
- Event bus, let's call it `EventBus`
- Our service, let's call it `DiscordService`
- Discord.NET REST API, let's call it `DiscordApi`

Interactions:

1. `EventBus` sends `MemberPromoted` event which gets consumed by `DiscordService`.
2. `DiscordService` authorizes itself to `DiscordApi` and obtains `DiscordRestClient`.
3. `DiscordService` maps old `RankId` to discord `RoleId`. It repeats the same for new rank. This is best held in settings.
4. `DiscordService` obtains `RestGuild` from `DiscordRestClient`, method `GetGuildAsync`. We will need `GuildId` as setting.
5. `DiscordService` obtains `RestGuildUser` from `RestGuild`, method `SearchUsersAsync`.
6. `DiscordService` calls `RemoveRoleAsync` on `RestGuildUser` to remove old role.
7. `DiscordService` calls `AddRoleAsync` on `RestGuildUser` to add new role.

**Reply by enrico-ghidoni** [reply]
Created [Discord Integration](https://github.com/CntoDev/rooster/milestone/2) milestone with the linked tasks

**Comment by milivojm** [comment]
Moved to General since it's under construction.

**Comment by enrico-ghidoni** [comment]
We'd be better off using the web socket part of the library since it's the most widely documented. I was unable to find any documentation except API Reference for the Rest part and there is no big difference when it comes to interfacing with it.

**Reply by milivojm** [reply]
API reference is all there is afaik for REST. Anyway, fine by me. You might as well create bot now. 💻

**Comment by enrico-ghidoni** [comment]
We have a small inconsistency issue with the `MemberPromoted` event still: the Discord member lookup is currently based on the member's `nickname` field, and this not only causes a possible match error in fetching the member on CNTO Discord (i.e. CNTO nickname and Discord name do not match, although it's rare) but also makes the `Discord ID` field in `Member` absolutely useless since the lookup doesn't use it. 

I suggest we simply update the `Discord ID` field into `Discord Nickname` to let applicants fill in their nickname on Discord (just the nickname, not the `#1234` code that follows). This would also require yet another update on `MemberPromoted` including the updated field.

@milivojm what do you think? We can make do with how it is right now and change it later, just something to be aware of.

**Reply by milivojm** [reply]
Let's do it now, it's a small change.

**Comment by milivojm** [comment]
We can close this, what do you think?

**Comment by enrico-ghidoni** [comment]
Well apparently there is no way to formally close a discussion, so I'm locking this
